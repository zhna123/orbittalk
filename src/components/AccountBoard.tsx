import { Outlet } from "react-router-dom";
import { Link } from "react-router-dom";

export function AccountBoard() {
  return (
    <>
    <div className="grid grid-cols-4 grow">
      <div className="bg-red-200 rounded-l-md">
        <ul className="text-xl text-grey-800">
          <Link to=''>
            <li className="hover:bg-red-300 rounded-tl-md p-6 text-red-800">
              <span>My Profile</span>
            </li>
          </Link>
          <Link to='security'>
            <li className="hover:bg-red-300 rounded-tl-md p-6 text-red-800">
              <span>Security</span>
            </li>
          </Link>
          <Link to='friends'>
            <li className="hover:bg-red-300 rounded-tl-md p-6 text-red-800">
              <span>Friends</span>
            </li>
          </Link>
          <Link to='notifications'>
            <li className="hover:bg-red-300 rounded-tl-md p-6 text-red-800">
              <span>Notifications</span>
            </li>
          </Link>
        </ul>
        <div className="h-0.5 bg-red-300 w-4/5 mx-auto mt-10"></div>
        <p className="p-6 text-red-600 font-medium text-lg">Sign Out</p>
      </div>
      <div className="col-span-3 bg-grey-100 rounded-r-md py-12 px-20">
        <Outlet />
      </div>
    </div>
    <hr className='h-0.5 text-grey-200 my-6 w-full mx-auto' />
    <footer className='text-grey-500 text-sm mt-auto'>
      &copy;2023 zhna123. All rights reserved.
    </footer>
    </>
  )
}