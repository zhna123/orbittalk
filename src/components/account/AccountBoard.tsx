import { useContext, useState } from "react";
import { Outlet } from "react-router-dom";
import { Link } from "react-router-dom";
import Icon from '@mdi/react';
import { mdiAccountDetails } from '@mdi/js';
import { mdiSecurity } from '@mdi/js';
import { mdiAccountGroupOutline } from '@mdi/js';
import { mdiBellOutline } from '@mdi/js';
import { mdiLogout } from '@mdi/js';
import { AppContext } from "../../App";


export function AccountBoard() {

  const { handleSignOut } = useContext(AppContext)

  const [selected, setSelected] = useState<string>('')

  return (
    <>
    <div className="flex grow">
      <div className="bg-red-500 rounded-l-md basis-20 sm:basis-52 shrink-0 grow-0">
        <p className="text-grey-100 text-sm sm:text-base p-2 sm:p-4 font-semibold">Account</p>
        <ul role="tablist" aria-label='account-settings-list' className="text-grey-100 text-xs sm:text-sm">
          <Link to=''>
            <li onClick={() => setSelected('profile')}
              role="tab" 
              aria-label="profile"
              aria-selected={selected === 'profile' ? 'true' : 'false'}
              tabIndex={selected === 'profile' ? 0 : -1}
              className={`hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 ${selected === 'profile' ? 'bg-red-400 bg-opacity-50 hover:bg-red-400 font-medium sm:font-semibold': ''}`}>
              <Icon path={mdiAccountDetails} size={1} className="hidden sm:block" />
              <span>My Profile</span>
            </li>
          </Link>
          <Link to='security'>
            <li onClick={() => setSelected('security')}
              role="tab" 
              aria-label="security"
              aria-selected={selected === 'security' ? 'true' : 'false'}
              tabIndex={selected === 'security' ? 0 : -1}
              className={`hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 ${selected === 'security' ? 'bg-red-400 bg-opacity-50 hover:bg-red-400 font-medium sm:font-semibold': ''}`}>
              <Icon path={mdiSecurity} size={1} className="hidden sm:block"/>
              <span>Security</span>
            </li>
          </Link>
          <Link to='friends'>
            <li onClick={() => setSelected('friends')}
              role="tab" 
              aria-label="friends"
              aria-selected={selected === 'friends' ? 'true' : 'false'}
              tabIndex={selected === 'friends' ? 0 : -1}
              className={`hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 ${selected === 'friends' ? 'bg-red-400 bg-opacity-50 hover:bg-red-400 font-medium sm:font-semibold': ''}`}>
              <Icon path={mdiAccountGroupOutline} size={1} className="hidden sm:block"/>
              <span>Friends</span>
            </li>
          </Link>
          <Link to='notifications'>
            <li onClick={() => setSelected('notifications')}
              role="tab" 
              aria-label="notifications"
              aria-selected={selected === 'notifications' ? 'true' : 'false'}
              tabIndex={selected === 'notifications' ? 0 : -1}
              className={`hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 ${selected === 'notifications' ? 'bg-red-400 bg-opacity-50 hover:bg-red-400 ffont-medium sm:ont-semibold': ''}`}>
              <Icon path={mdiBellOutline} size={1} className="hidden sm:block"/>
              <span>Notifications</span>
            </li>
          </Link>
        </ul>
        <div className="h-0.5 bg-red-300 w-4/5 mx-auto mt-8"></div>
        <div className="flex gap-4 items-center p-2 mt-2 sm:p-4 text-grey-100 cursor-pointer">
          <Icon path={mdiLogout} size={1} className="hidden sm:block"/>
          <p className="text-xs sm:text-sm" onClick={handleSignOut}>Sign Out</p>
        </div>
      </div>
      <div className="bg-grey-100 rounded-r-md p-4 grow sm:py-12 sm:px-20">
        <Outlet />
      </div>
    </div>
    <hr className='h-0.5 text-grey-200 my-6 w-full mx-auto' />
    <footer className='text-grey-500 text-sm mt-auto'>
      &copy;2023 zhna123. All rights reserved.
    </footer>
    </>
  )
}