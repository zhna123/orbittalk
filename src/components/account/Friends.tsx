import Icon from '@mdi/react';
import { mdiTrashCan } from '@mdi/js';
import { mdiPlus } from '@mdi/js';
import { mdiCircle } from '@mdi/js';
import AVATAR_DEFAULT from '../../assets/avatar-default.png'
import { useFriendList } from '../../hooks/useFriendList';
import { useState } from 'react';
import { AddFriendModule } from './AddFriendModule';
import { User } from '../../types';
import { SERVER_URL } from '../../util/constant';




export function Friends() {

  const { isLoading, friendList, addFriend } = useFriendList()

  const [openModule, setOpenModule] = useState<boolean>(false)

  const handleAddFriend = async (friend: User) => {
    const res = await fetch(`${SERVER_URL}/users/friend`, {
      method: "PUT",
      mode: "cors",
      credentials: "include",
      body: JSON.stringify({"friendId": friend._id}),
      headers: {
        'Content-Type': 'application/json',
      }
    })
    if (res.status === 200) {
      console.log("friend added")
      const result = await res.json()
      addFriend(result.data)
      setOpenModule(false)
    } else {
      console.log("failed to add friend")
    }
  }

  return (
    <>
      <div className='flex gap-4 items-center mb-6 sm:mb-10'>
        <h1 className='text-lg sm:text-xl text-grey-800 font-semibold'>My Friends</h1>
        <div className='relative hover:cursor-pointer' onClick={() => setOpenModule(true) }>
          <Icon path={mdiPlus} className='text-grey-800 absolute top-1/4 left-1/4 h-5 sm:h-6' />
          <Icon path={mdiCircle} className='text-red-200 h-10 sm:h-12' />
        </div>
      </div>
      <ul className='max-w-md flex flex-col gap-2 text-xs sm:text-sm max-h-full overflow-auto'>
          {
            isLoading ? <li className='text-grey-800'>Loading friends...</li> :
            (
              friendList.length === 0 ? <li>No friend added yet.</li> :
              <>
              { friendList.map(f => (
                  <div key={f!._id}>
                    <li className='flex justify-between items-center'>
                      <div className='flex items-center gap-2 sm:gap-4'>
                        <img src={ AVATAR_DEFAULT } alt="avatar" className='h-6 sm:h-10' />
                        <span className='text-grey-800'>{f!.screen_name}</span>
                      </div>
                      <div className='relative hover:cursor-pointer'>
                        <Icon path={mdiTrashCan} className='text-grey-800 absolute top-1/4 left-1/4 h-4 sm:h-5' />
                        <Icon path={mdiCircle} className='text-red-200 h-8 sm:h-10' />
                      </div>
                    </li>
                    <div className='h-0.5 bg-red-200 mt-1 sm:mt-2'></div>
                  </div>
                )) 
              }
              </>
            )
          }
      </ul>
      {
        openModule && <AddFriendModule openModule={(b: boolean) => setOpenModule(b)} handleAddFriend={handleAddFriend} />
      }
    </>
  )
}