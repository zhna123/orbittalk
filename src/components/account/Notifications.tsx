import AVATAR_DEFAULT from '../../assets/avatar-default.png'

export function Notifications() {
  return (
    <>
      <h1 className="text-grey-800 text-lg sm:text-xl mb-6 sm:mb-10 font-semibold">Notifications</h1>
      <ul>
        <li className='flex gap-2 sm:gap-4 text-sm sm:text-lg'>
          <img src={ AVATAR_DEFAULT } alt="avatar" className='h-6 sm:h-8' />
          <div>
            <p className='text-grey-800 text-xs sm:text-sm'><span className='italic'>April</span> wants to add you as a friend.</p>
            <p className='text-grey-500 mb-1 sm:mb-2 text-xs sm:text-sm'>5 mins ago</p>
            <button className='bg-red-500 sm:px-3 px-2 sm:py-1 mr-4 rounded-md text-red-100 text-[10px] sm:text-xs'>Accept</button>
            <button className='outline outline-grey-600 px-2 sm:px-3 sm:py-1 rounded-md text-grey-700 text-[10px] sm:text-xs'>Ignore</button>
          </div>
        </li>
        <div className='h-0.5 bg-red-200 my-2 sm:my-4'></div>
      </ul>
    </>
  )
}