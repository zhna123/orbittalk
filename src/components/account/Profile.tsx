import Icon from '@mdi/react';
import { mdiPencil } from '@mdi/js';
import { mdiCircle } from '@mdi/js';
import AVATAR_DEFAULT from '../../assets/avatar-default.png'
import { ChangeEvent, useContext, useEffect, useState } from 'react';
import { AppContext } from '../../App';
import { getAvatarDataUri } from '../../util/image';
import { SERVER_URL } from '../../util/constant';




export function Profile() {

  const { user, updateUserAvatar } = useContext(AppContext)

  const [selectedFile, setSelectedFile] = useState<File>()
  const [preview, setPreview] = useState<string>()

  // create a preview as a side effect, whenever selected file is changed
  useEffect(() => {
    if (!selectedFile) {
        setPreview(undefined)
        return
    }
    const objectUrl = URL.createObjectURL(selectedFile)
    setPreview(objectUrl)

    // free memory when ever this component is unmounted
    return () => URL.revokeObjectURL(objectUrl)
  }, [selectedFile])

  const onSelectFile = (e: ChangeEvent<HTMLInputElement>) => {

    if (!e.target.files || e.target.files.length === 0) {
        setSelectedFile(undefined)
        return
    }
    setSelectedFile(e.target.files[0])
  }

  const onUpload = async () => {
    const formData = new FormData()
    formData.append("avatar", selectedFile!)
    
    // upload avatar
    const res = await fetch(`${SERVER_URL}/users/avatar`, {
      method: "PUT",
      mode: "cors",
      credentials: "include",
      body: formData,
    })
    const result = await res.json()
    if (res.status === 200) {
      console.log("upload succeed")
      updateUserAvatar(result.data)
      setSelectedFile(undefined)
    } else {
      console.log("upload error:" + result.error)
    }
  }

  const displayAvatar = () => {
    if (user!.avatar && !selectedFile) {
      return <img src={ getAvatarDataUri(user!.avatar) } alt="profile photo" className='h-24 w-24 sm:h-44 sm:w-44 rounded-full object-cover' />
    } else {
      if (selectedFile) {
        return <img src={preview} alt="profile photo" className='h-24 w-24 sm:h-44 sm:w-44 rounded-full object-cover' /> 
      }
      return <img src={AVATAR_DEFAULT} alt="profile photo" className='h-24 w-24 sm:h-44 sm:w-44 rounded-full object-cover' />
    }
  }

  return (
    <>
      <h1 className="text-lg sm:text-xl mb-6 sm:mb-10 text-grey-800 font-semibold">My Profile</h1>
      <div className='flex sm:flex-row flex-col gap-8'>
        <div className='flex flex-col gap-6 shrink-0'>
          { 
            displayAvatar()
          }
          <div className='flex flex-col gap-4'>
            {/* commented out because we can't use keydown on the label element / need to find a better way to style this */}
            {/* <label htmlFor='avatar' role="button" aria-label="Choose an image" 
                    className='sm:self-center underline text-grey-800'
                    tabIndex={0}>
              Click to select an image
              <input type="file" id='avatar' name='avatar'
                accept='.png, .jpeg, .jpg, .gif'
                onChange={ onSelectFile }
                className='hidden'/>
            </label> */}
            <input aria-label='choose avatar' type="file" id='avatar' name='avatar'
                accept='.png, .jpeg, .jpg, .gif'
                onChange={ onSelectFile } />
            {
              selectedFile ?
                <button aria-label='upload' className='bg-red-500 text-red-100 p-2 rounded-md font-medium text-xs sm:text-sm w-1/2' onClick={onUpload}>
                Upload
                </button>
                :
                <button aria-label='upload-disabled' className='bg-grey-500 text-grey-100 p-2 rounded-md font-medium text-xs sm:text-sm w-1/2 disabled'>
                Upload
                </button>
            }
            <ul className='text-red-600 text-xs'>
              <li>* Accept jpg, jpeg, gif, png file format</li>
              <li>* File size needs to be less than 300KB</li>
            </ul>
          </div>
        </div>
        {/* need to specify min-w-0 because the flex item will be at least the width of the narrowest child, 
          then break-words won't work */}
        <div className='text-grey-800 min-w-0'>
          <p className='text-lg sm:text-xl my-6'>{user!.username}</p>
          
          <div className='flex gap-4 items-center mb-4 text-sm sm:text-lg'>
            <p id='screen name'>{user!.screen_name}</p>
            <button aria-labelledby='screen name' className='relative hover:cursor-pointer'>
              <Icon path={mdiPencil} size={.7} className='text-grey-800 absolute top-1/4 left-1/4' />
              <Icon path={mdiCircle} size={1.4} className='text-red-200' />
            </button>
          </div>

          <p className='break-words'>{user!.email}</p>
        </div>
      </div>
     
    </>
  )
}