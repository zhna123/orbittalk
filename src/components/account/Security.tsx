import { useForm, SubmitHandler } from "react-hook-form"
import { ErrorMessage } from "@hookform/error-message"
import { useState } from "react";
import { SERVER_URL } from "../../util/constant";



type Inputs = {
  oldPassword: string,
  password: string,
  passwordConfirmation: string
}

interface ErrorDivProps {
  name: keyof Inputs;
}

interface FormErrorType {
  msg: string
}

export function Security() {

  const [success, setSuccess] = useState<boolean>(false)
  const [formErrors, setFormErrors] = useState<FormErrorType[]>([])

  const {
    register,
    watch,
    reset,
    formState: { errors },
    handleSubmit,
  } = useForm<Inputs>({
    criteriaMode: "all",
    defaultValues: {
      oldPassword: '',
      password: '',
      passwordConfirmation: ''
    }
  })

  const ErrorDiv = ({name}: ErrorDivProps) => {
    return (
      <ErrorMessage
            errors={errors}
            name={name}
            render={({ messages }) =>
              messages &&
              Object.entries(messages).map(([type, message]) => (
                <p key={type} role="alert" className="text-red-600 text-sm font-medium">{message}</p>
              ))
            }
      />
    )
  }

  const onSubmit: SubmitHandler<Inputs> = async (data) => {
    try {
      const res = await fetch(`${SERVER_URL}/users/password`, {
        method: "PUT",
        mode: "cors",
        credentials: "include",
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json',
        }
      })
      if (res.status === 200) {
        console.log('password changed!')
        setSuccess(true)
        setFormErrors([])
        reset()
      } else {
        console.log('failed change pwd')
        const msg = await res.json()
        setSuccess(false)
        setFormErrors(msg)
        reset()
      }
    } catch(e) {
      console.log('error when changing password:' + e)
    }
  }
 
  return (
    <>
      <p className="text-lg sm:text-xl text-grey-800 mb-6 sm:mb-10 font-semibold">Change Password</p>
      <form onSubmit={ handleSubmit(onSubmit) } className="flex flex-col gap-4 max-w-xs text-xs sm:text-sm">
        <label htmlFor="old_pwd" className="text-grey-800">Old Password <span className="text-red-600 ">*</span></label>
        <input type="password" id="old_pwd" autoComplete="true"
          aria-invalid={errors.password ? "true" : "false"}
          {...register("oldPassword", {
            required: "This is required",
          })}
          className={`form-input rounded-md border-grey-300 ${errors.oldPassword ? 'focus:ring-0 focus:border-red-500'
            :'focus:ring-0 focus:border-grey-600 focus:shadow-inner' }`} />
        <ErrorDiv name="oldPassword" />

        <label htmlFor="password" className="text-grey-800">New Password <span className="text-red-600 ">*</span></label>
        <input type="password" id="password" autoComplete="true"
          aria-invalid={errors.password ? "true" : "false"}
          {...register("password", {
            required: "This is required",
            minLength: {
              value: 8,
              message: "Password must be at least 8 characters",
            },
          })}
          className={`form-input rounded-md border-grey-300 ${errors.password ? 'focus:ring-0 focus:border-red-500'
            :'focus:ring-0 focus:border-grey-600 focus:shadow-inner' }`} />
        <ErrorDiv name="password" />

        <label htmlFor="confirm_pwd" className="text-grey-800">Confirm New Password <span className="text-red-600 ">*</span></label>
        <input type="password" id="confirm_pwd" autoComplete="true"
          aria-invalid={errors.passwordConfirmation ? "true" : "false"}
          {...register("passwordConfirmation", {
            required: "This is required",
            validate: (val: string) => {
              if (watch("password") !== val) {
                return "Passwords do not match"
              }
            }
          })}
          className={`form-input rounded-md border-grey-300 ${errors.passwordConfirmation ? 'focus:ring-0 focus:border-red-500'
            :'focus:ring-0 focus:border-grey-600 focus:shadow-inner' }`} />
        <ErrorDiv name="passwordConfirmation" />

        { success ? <p className="text-green">New password saved successfully!</p>
          : 
          <ul className="text-red-600 font-medium text-sm">
            { formErrors.map(err => <li>{err.msg}</li>)}
          </ul>
        }

        <input type="submit" value='Save' className="form-input text-xs sm:text-sm font-medium mt-2 sm:mt-4 border-none rounded-md text-red-100 bg-red-500 cursor-pointer focus:ring-0 active:bg-red-600" />
      </form>
      
    </>
  )
}