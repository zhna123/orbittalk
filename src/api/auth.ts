import { SERVER_URL } from "../util/constant"
import { Inputs } from "../components/SignIn"


export const signInUser = async (data: Inputs) => {
    
  try {
    const res = await fetch(`${SERVER_URL}/auth/login`, {
      method: "POST",
      mode: "cors",
      credentials: "include",
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json',
      }
    })
    if (res.status === 201) {
      console.log("log in succeed.")
      return true
    } else {
      console.log("log in failed.")
      return false
    }
  } catch (e) {
    console.log("error when log in:" + e)
    return false
  }
}