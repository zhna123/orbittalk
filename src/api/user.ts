import { redirect } from "react-router-dom"
import { Inputs } from "../components/Register"
import { User } from "../types"
import { SERVER_URL } from "../util/constant"

export const registerUser = async (data: Inputs) => {
  try {
    const res = await fetch(`${SERVER_URL}/sign-up`, {
      method: "POST",
      mode: "cors",
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json',
      }
    })
    const result = await res.json()
    if (res.status === 201) {
      console.log("user registered successful")
      return [true, result]
    } else {
      console.log("user failed to register")
      return [false, result]
    }
  } catch (e) {
    console.log("error when register:" + e)
    return [false, {}]
  }
}

export const findFriends = async (user: User) => {

  const friendIds: string[] = user.friends;

  const friendsPromises: Promise<User | null>[] = friendIds.map(async id => {
    const res = await fetch(`${SERVER_URL}/users/${id}`, {
      method: "GET",
      mode: "cors",
      credentials: "include",
      headers: {
        'Content-Type': 'application/json',
      }
    })
    if (res.status === 200) {
      const friend: User = await res.json()
      return friend;
    } else if (res.status === 401 || res.status === 403) {
      redirect('/')
      return null
    } else {
      console.log(`Failed to retrieve friend data for user ID: ${id}`);
      return null
    }
  })
  const friendsData = await Promise.all(friendsPromises)
  const filteredData = friendsData.filter(data => data !== null);
  return filteredData
}