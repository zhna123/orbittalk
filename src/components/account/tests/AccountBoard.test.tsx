import { render, screen } from "@testing-library/react";
import { AccountBoard } from "../AccountBoard";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import {
  RouterProvider,
  createMemoryRouter,
} from "react-router-dom";
import { Security } from "../Security";
import { Friends } from "../Friends";
import { Notifications } from "../Notifications";
import {vi} from "vitest"
import * as user from "../../../api/user"
import { Profile } from "../Profile";

vi.spyOn(user, "findFriends").mockImplementation(async() => [])

vi.mock('../Profile', () => ({ 
  Profile: () => {
    return <div data-testid="profile">My Profile</div>
  }, 
})) 

describe("account board", () => {
  it("renders account board", () => {
    const { container } = render(<BrowserRouter>
        <AccountBoard />
      </BrowserRouter>)
    expect(container).toMatchSnapshot()
  })
  it("should show profile by default", async () => {
    const routes = [
      {
        path: "/account",
        element: <AccountBoard />,
      },
    ];
    const router = createMemoryRouter(routes, {
      initialEntries: ["/account"],
      initialIndex: 1,
    });
    render(<RouterProvider router={router} />);
    expect(screen.getByText("My Profile")).toBeInTheDocument()
  })
  it("should show correct tab content when user clicks on the tab navigation", async () => {
    const routes = [
      {
        path: "/account",
        element: <AccountBoard />,
        children: [
          {
            path: "",
            element:  <Profile />
          },
          {
            path: "security",
            element: <Security />
          },
          {
            path: "friends",
            element: <Friends />
          },
          {
            path: "notifications",
            element: <Notifications />
          }
        ]
      },
      
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ["/account"],
      initialIndex: 1,
    });
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    const tab_profile = screen.getByRole("tab", {name: "profile"})
    await user.click(tab_profile)
    expect(screen.getByTestId("profile").textContent).toMatch("My Profile")
    
    const tab_security = screen.getByRole("tab", {name: "security"})
    await user.click(tab_security)
    expect(screen.getByRole("heading").textContent).toMatch("Change Password")

    const tab_friends = screen.getByRole("tab", {name: "friends"})
    await user.click(tab_friends)
    expect(screen.getByRole("heading").textContent).toMatch("My Friends")

    const tab_notify = screen.getByRole("tab", {name: "notifications"})
    await user.click(tab_notify)
    expect(screen.getByRole("heading").textContent).toMatch("Notifications")
  })
})