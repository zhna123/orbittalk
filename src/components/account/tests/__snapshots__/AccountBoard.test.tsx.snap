// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account board > renders account board 1`] = `
<div>
  <div
    class="flex grow"
  >
    <div
      class="bg-red-500 rounded-l-md basis-20 sm:basis-52 shrink-0 grow-0"
    >
      <p
        class="text-grey-100 text-sm sm:text-base p-2 sm:p-4 font-semibold"
      >
        Account
      </p>
      <ul
        aria-label="account-settings-list"
        class="text-grey-100 text-xs sm:text-sm"
        role="tablist"
      >
        <a
          href="/"
        >
          <li
            aria-label="profile"
            aria-selected="false"
            class="hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 "
            role="tab"
            tabindex="-1"
          >
            <svg
              class="hidden sm:block"
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 9C11 10.66 9.66 12 8 12C6.34 12 5 10.66 5 9C5 7.34 6.34 6 8 6C9.66 6 11 7.34 11 9M14 20H2V18C2 15.79 4.69 14 8 14C11.31 14 14 15.79 14 18M22 12V14H13V12M22 8V10H13V8M22 4V6H13V4Z"
                style="fill: currentColor;"
              />
            </svg>
            <span>
              My Profile
            </span>
          </li>
        </a>
        <a
          href="/security"
        >
          <li
            aria-label="security"
            aria-selected="false"
            class="hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 "
            role="tab"
            tabindex="-1"
          >
            <svg
              class="hidden sm:block"
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,12H19C18.47,16.11 15.72,19.78 12,20.92V12H5V6.3L12,3.19M12,1L3,5V11C3,16.55 6.84,21.73 12,23C17.16,21.73 21,16.55 21,11V5L12,1Z"
                style="fill: currentColor;"
              />
            </svg>
            <span>
              Security
            </span>
          </li>
        </a>
        <a
          href="/friends"
        >
          <li
            aria-label="friends"
            aria-selected="false"
            class="hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 "
            role="tab"
            tabindex="-1"
          >
            <svg
              class="hidden sm:block"
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,5A3.5,3.5 0 0,0 8.5,8.5A3.5,3.5 0 0,0 12,12A3.5,3.5 0 0,0 15.5,8.5A3.5,3.5 0 0,0 12,5M12,7A1.5,1.5 0 0,1 13.5,8.5A1.5,1.5 0 0,1 12,10A1.5,1.5 0 0,1 10.5,8.5A1.5,1.5 0 0,1 12,7M5.5,8A2.5,2.5 0 0,0 3,10.5C3,11.44 3.53,12.25 4.29,12.68C4.65,12.88 5.06,13 5.5,13C5.94,13 6.35,12.88 6.71,12.68C7.08,12.47 7.39,12.17 7.62,11.81C6.89,10.86 6.5,9.7 6.5,8.5C6.5,8.41 6.5,8.31 6.5,8.22C6.2,8.08 5.86,8 5.5,8M18.5,8C18.14,8 17.8,8.08 17.5,8.22C17.5,8.31 17.5,8.41 17.5,8.5C17.5,9.7 17.11,10.86 16.38,11.81C16.5,12 16.63,12.15 16.78,12.3C16.94,12.45 17.1,12.58 17.29,12.68C17.65,12.88 18.06,13 18.5,13C18.94,13 19.35,12.88 19.71,12.68C20.47,12.25 21,11.44 21,10.5A2.5,2.5 0 0,0 18.5,8M12,14C9.66,14 5,15.17 5,17.5V19H19V17.5C19,15.17 14.34,14 12,14M4.71,14.55C2.78,14.78 0,15.76 0,17.5V19H3V17.07C3,16.06 3.69,15.22 4.71,14.55M19.29,14.55C20.31,15.22 21,16.06 21,17.07V19H24V17.5C24,15.76 21.22,14.78 19.29,14.55M12,16C13.53,16 15.24,16.5 16.23,17H7.77C8.76,16.5 10.47,16 12,16Z"
                style="fill: currentColor;"
              />
            </svg>
            <span>
              Friends
            </span>
          </li>
        </a>
        <a
          href="/notifications"
        >
          <li
            aria-label="notifications"
            aria-selected="false"
            class="hover:bg-red-300 hover:bg-opacity-50 flex gap-4 items-center p-2 sm:p-4 "
            role="tab"
            tabindex="-1"
          >
            <svg
              class="hidden sm:block"
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 21H14C14 22.1 13.1 23 12 23S10 22.1 10 21M21 19V20H3V19L5 17V11C5 7.9 7 5.2 10 4.3V4C10 2.9 10.9 2 12 2S14 2.9 14 4V4.3C17 5.2 19 7.9 19 11V17L21 19M17 11C17 8.2 14.8 6 12 6S7 8.2 7 11V18H17V11Z"
                style="fill: currentColor;"
              />
            </svg>
            <span>
              Notifications
            </span>
          </li>
        </a>
      </ul>
      <div
        class="h-0.5 bg-red-300 w-4/5 mx-auto mt-8"
      />
      <div
        class="flex gap-4 items-center p-2 mt-2 sm:p-4 text-grey-100 cursor-pointer"
      >
        <svg
          class="hidden sm:block"
          role="presentation"
          style="width: 1.5rem; height: 1.5rem;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16,17V14H9V10H16V7L21,12L16,17M14,2A2,2 0 0,1 16,4V6H14V4H5V20H14V18H16V20A2,2 0 0,1 14,22H5A2,2 0 0,1 3,20V4A2,2 0 0,1 5,2H14Z"
            style="fill: currentColor;"
          />
        </svg>
        <p
          class="text-xs sm:text-sm"
        >
          Sign Out
        </p>
      </div>
    </div>
    <div
      class="bg-grey-100 rounded-r-md p-4 grow sm:py-12 sm:px-20"
    />
  </div>
  <hr
    class="h-0.5 text-grey-200 my-6 w-full mx-auto"
  />
  <footer
    class="text-grey-500 text-sm mt-auto"
  >
    Â©2023 zhna123. All rights reserved.
  </footer>
</div>
`;
